lib/smack/DSAWrapper.cpp

```diff
@@ -22,185 +25,70 @@ RegisterPass<DSAWrapper>
     DSAWrapperPass("dsa-wrapper",
                    "SMACK Data Structure Graph Based Alias Analysis Wrapper");
 
-void MemcpyCollector::visitMemCpyInst(llvm::MemCpyInst &mci) {
-  const llvm::EquivalenceClasses<const llvm::DSNode *> &eqs =
-      nodeEqs->getEquivalenceClasses();
-  const llvm::DSNode *n1 =
-      eqs.getLeaderValue(nodeEqs->getMemberForValue(mci.getOperand(0)));
-  const llvm::DSNode *n2 =
-      eqs.getLeaderValue(nodeEqs->getMemberForValue(mci.getOperand(1)));
-
-  bool f1 = false, f2 = false;
-  for (unsigned i = 0; i < memcpys.size() && (!f1 || !f2); i++) {
-    f1 = f1 || memcpys[i] == n1;
-    f2 = f2 || memcpys[i] == n2;
-  }
-
-  if (!f1)
-    memcpys.push_back(eqs.getLeaderValue(n1));
-  if (!f2)
-    memcpys.push_back(eqs.getLeaderValue(n2));
-}
-
 void DSAWrapper::getAnalysisUsage(llvm::AnalysisUsage &AU) const {
   AU.setPreservesAll();
-  AU.addRequiredTransitive<llvm::BUDataStructures>();
-  AU.addRequiredTransitive<llvm::TDDataStructures>();
-  AU.addRequiredTransitive<llvm::DSNodeEquivs>();
-  AU.addRequired<dsa::TypeSafety<llvm::TDDataStructures>>();
+  AU.addRequiredTransitive<sea_dsa::DsaAnalysis>();
 }
 
 bool DSAWrapper::runOnModule(llvm::Module &M) {
   dataLayout = &M.getDataLayout();
-  TD = &getAnalysis<llvm::TDDataStructures>();
-  BU = &getAnalysis<llvm::BUDataStructures>();
-  nodeEqs = &getAnalysis<llvm::DSNodeEquivs>();
-  TS = &getAnalysis<dsa::TypeSafety<llvm::TDDataStructures>>();
-  memcpys = collectMemcpys(M, new MemcpyCollector(nodeEqs));
-  staticInits = collectStaticInits(M);
+  SD = &getAnalysis<sea_dsa::DsaAnalysis>().getDsaAnalysis();
+  assert(SD->kind() == sea_dsa::GlobalAnalysisKind::CONTEXT_INSENSITIVE &&
+         "Currently we only want the context-insensitive sea-dsa.");
+  DG = &SD->getGraph(*M.begin());
+  collectStaticInits(M);
+  collectMemOpds(M);
+  countGlobalRefs();
   module = &M;
   return false;
 }
 
-std::vector<const llvm::DSNode *>
-DSAWrapper::collectMemcpys(llvm::Module &M, MemcpyCollector *mcc) {
-
-  for (llvm::Module::iterator func = M.begin(), e = M.end(); func != e;
-       ++func) {
-
-    for (llvm::Function::iterator block = func->begin(); block != func->end();
-         ++block) {
-
-      mcc->visit(*block);
-    }
-  }
-  return mcc->getMemcpys();
-}
-
-std::vector<const llvm::DSNode *>
-DSAWrapper::collectStaticInits(llvm::Module &M) {
-  std::vector<const llvm::DSNode *> sis;
+void DSAWrapper::collectStaticInits(llvm::Module &M) {
   for (GlobalVariable &GV : M.globals()) {
     if (GV.hasInitializer()) {
       if (auto *N = getNode(&GV)) {
```

Comment:
Can this ever be null?

Reply:
I'm not sure. I would say it shouldn't be null here. Should I add an assertion?

Reply:
I guess just add it. I cannot hurt.